### DISCUSSION TITLE
a suggestion to avoid ClassCircularityError

### DISCUSSION THREAD

**Discussion by shichaoyuan** [apache/skywalking#11331] [General]
Due to the use of `ThreadLocalRandom` in my custom plugin, `ClassCircularityError` are occasionally triggered.

I found that many agents have related issues:
1. https://github.com/DataDog/dd-trace-java/pull/4865
2. https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/3315
3. https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/4557
4. https://github.com/newrelic/newrelic-java-agent/pull/600
...

Since we will use many jdk classes in custom plugins，should we ignore most `java.*` classes？

![image](https://github.com/apache/skywalking/assets/1159482/2f1eecdb-ba8a-440f-8564-3362d3e57010)

**Comment by wu-sheng** [comment]
Why do we need to worry about something on other projects' sides? I am a little confused.

**Comment by shichaoyuan** [comment]
Because `ClassCircularityError` may also be encountered in Skywalking’s agent, if certain jdk classes are used in the plugin.

**Reply by wu-sheng** [reply]
But instrument JDK class is acceptable, that is why bootstrap plugins exist.

**Reply by shichaoyuan** [reply]
I think we can ignore most `java.*` classes, only allow some in bootstrap plugins.

**Reply by wu-sheng** [reply]
I don't think this is a big issue. For the worst case, a bootstrap plugin can't be used, which should be tested during contribution review stage.
But excluding them would cause a lot of questions, I don't think I should suffer that noise.

**Reply by shichaoyuan** [reply]
refer to https://bugs.openjdk.org/browse/JDK-8164165

![image](https://github.com/apache/skywalking/assets/1159482/f7d780dd-ecec-4da1-a30b-ee4e6dfb45fc)

Class loading may be triggered by user code, which may cause user code to throw ClassCircularityError.

**Reply by wu-sheng** [reply]
Yes. This statement is correct for sure, and your scenario could happen too. Don't get me wrong, I am not arguing this.

My point is, bootstrap plugin is not activated by default, so, we would not face this by default.
So, that error case should only happen when a bootstrap plugin is acrivated. This is not a big concern for a project or end users, who are adding them by themselves.
But, if we accept your proposals, developers would be very confused why we block things we are not using. 

Is this making you more clear?

**Reply by shichaoyuan** [reply]
Got it, the end user is responsible

**Reply by wu-sheng** [reply]
We as maintainer would try our best to keep codes as good as possible, to avoid this conflict. But no more on codes level.
